### PR TITLE
PRT-214 Refactor Swap methods

### PR DESCRIPTION
**Pull request by lazaralex98**
**Important:** #26 is built on top of this, so it may be a good idea to review this first?

I moved quickly on [PRT-212](https://linear.app/toucan/issue/PRT-212/calculate-token-amount-to-send-method) so I could keep on working on the SDK, so I've left somewhat of a mess that needed to be cleaned up.

### CODE REVIEW

File: contracts/OffsetHelper.sol [hunk 1]
```diff
@@ -280,17 +254,17 @@ contract OffsetHelper is OffsetHelperStorage {
...
-            "You don't have enough to withdraw."
+            "Insufficient balance"
         );
 
-        IERC20(_erc20Addr).transfer(msg.sender, _amount);
+        IERC20(_erc20Addr).safeTransfer(msg.sender, _amount);
         balances[msg.sender][_erc20Addr] -= _amount;
     }
 
     // @description allow people to deposit BCT / NCT
```
Comment (kargakis) [inline contracts/OffsetHelper.sol:264]: This comment is not correct anymore.

**Comment by kargakis** [comment]
It's not clear from PRT-214 description what is the goal of this refactor. Personally I am in favor of keeping the more special-cased `is*` functions as the single `isEligible` function enables using more tokens where users shouldn't be able to do so.

File: contracts/OffsetHelper.sol (hunk 1 above)
Comment (lazaralex98) [reply contracts/OffsetHelper.sol:264]: What do you mean?

Comment (kargakis) [reply contracts/OffsetHelper.sol:264]: `isRedeemable` allowed the use of only NCT or BCT whereas now `isEligible` allows the use of a bunch of other tokens on top, right? I think this is problematic in general, see https://github.com/ToucanProtocol/example-implementations/pull/25#issuecomment-1147239552

**Comment by lazaralex98** [comment]
> 

The initial goal of the refactor was to clean up the swap methods as they were inconsistent in style to each other and quite messy (one of them used its calculation method, the other didn't, etc). These were things you noticed too in #23 , but I said I'll open an issue for it because I wanted to focus on the SDK and #23 solved an issue that was blocking me. 

Refactoring those, I realised I could make some gas optimisations. I did those. 

Also, I thought that while using the more special-cased `is*` is indeed somewhat safer, most of the time it checks for things needlessly:
- if someone tries to redeem a USDC it will fail anyway
- if someone tries to swap USDC for WETH, they can swap back
- there is a `withdraw` function in case that is needed

Perhaps, **most importantly, users are not even supposed to use the individual `swap`, `autoRedeem` or `autoRetire` methods.** They are supposed to use the `autoOffset` methods which, by design, will not try to redeem wrong addresses, etc.

**Comment by lazaralex98** [comment]
I'm open to bringing the special-cased methods back if you think it's important @kargakis , but that's just my argument against them

**Comment by kargakis** [comment]
> Refactoring those, I realised I could make some gas optimisations. I did those.
> 
> Also, I thought that while using the more special-cased `is*` is indeed somewhat safer, most of the time it checks for things needlessly:

It would be useful to track gas snapshots because now some of these functions are more expensive than before due to the use of `isEligible` which does a bunch of checks in place (including an external contract call) vs the previous functions.

**Comment by lazaralex98** [comment]
Gas snapshot after PRT-215:
```
·---------------------------------------------|---------------------------|-------------|-----------------------------·
|             Solc version: 0.8.4             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
··············································|···························|·············|······························
|  Methods                                                                                                            │
·················|····························|·············|·············|·············|···············|··············
|  Contract      ·  Method                    ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingETH        ·          -  ·          -  ·     483037  ·            1  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingPoolToken  ·          -  ·          -  ·     327984  ·            1  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingToken      ·     450446  ·     495827  ·     482943  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoRedeem                ·     157007  ·     159234  ·     158121  ·            6  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoRetire                ·          -  ·          -  ·     241320  ·            2  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  deposit                   ·     103380  ·     105585  ·     104758  ·           16  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  swap                      ·     263046  ·     277627  ·     270337  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  swap                      ·     267398  ·     269864  ·     268631  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  withdraw                  ·          -  ·          -  ·      42213  ·            2  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  Swapper       ·  swap                      ·     161067  ·     220743  ·     203884  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  Deployments                                ·                                         ·  % of limit   ·             │
··············································|·············|·············|·············|···············|··············
|  OffsetHelper                               ·          -  ·          -  ·    2206800  ·        7.4 %  ·          -  │
··············································|·············|·············|·············|···············|··············
|  Swapper                                    ·          -  ·          -  ·     733146  ·        2.4 %  ·          -  │
·---------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

Gas snapshot before PRT-214:
```
·---------------------------------------------|---------------------------|-------------|-----------------------------·
|             Solc version: 0.8.4             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
··············································|···························|·············|······························
|  Methods                                                                                                            │
·················|····························|·············|·············|·············|···············|··············
|  Contract      ·  Method                    ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingETH        ·          -  ·          -  ·     475564  ·            1  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingPoolToken  ·          -  ·          -  ·     322190  ·            1  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoOffsetUsingToken      ·     496056  ·     498012  ·     497034  ·            2  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoRedeem                ·     151397  ·     153625  ·     152511  ·            6  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  autoRetire                ·          -  ·          -  ·     241330  ·            2  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  deposit                   ·      90736  ·      92942  ·      92115  ·           16  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  swap                      ·     250172  ·     264753  ·     257463  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  swap                      ·     264988  ·     267194  ·     266091  ·            4  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  OffsetHelper  ·  withdraw                  ·          -  ·          -  ·      41140  ·            2  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  Swapper       ·  swap                      ·     212722  ·     220405  ·     217818  ·            3  ·          -  │
·················|····························|·············|·············|·············|···············|··············
|  Deployments                                ·                                         ·  % of limit   ·             │
··············································|·············|·············|·············|···············|··············
|  OffsetHelper                               ·          -  ·          -  ·    2345387  ·        7.8 %  ·          -  │
··············································|·············|·············|·············|···············|··············
|  Swapper                                    ·          -  ·          -  ·     737250  ·        2.5 %  ·          -  │
·---------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

**Pull request merged**